### DISCUSSION TITLE
Custom Pilots

### DISCUSSION THREAD

**Discussion by janfromme** [NVlabs/sionna#36] [Q&A] [answered]
Hi everyone,

the custom pilot pattern mask lets me configure at which OFDM Symbol location I would like to add a pilot symbol. That works pretty well, but what I cannot do currently is to also specify the subcarrier indexes that I would like to use for the pilot transmission. Sionna apparently assumes that the pilots are transmitted on all the subcarriers. However, in 5G NR the DM-RS locations are specified by OFDM symbol indexes as well as subcarrier index.

I'm thinking of adding that feature but wanted to check with you first in case I got something completely wrong here. What do you think?

(oh and great project btw.)

Regards,
Jan

**Comment by jhoydis** [comment, marked as answer]
Hi Jan,

You can actually define arbitrary pilot patterns over a resource grid. Please check the API documentation https://nvlabs.github.io/sionna/api/ofdm.html#pilot-pattern. 

The ``mask``, which defines which resource elements are reserved for pilots, has shape ``[num_tx, num_streams_per_tx, num_ofdm_symbols, num_effective_subcarriers]``. Thus, you can define for every stream of every transmitter an arbitrary pattern consisting of subcarrier and OFDM symbol indices.

Note that the ``mask`` only defines which resource elements are reserved for pilots. The ``pilots`` tensor of shape ``[num_tx, num_txt_ant, num_pilots]`` then defines what is actually transmitted on these reserved resource elements. 

You can see in the example provided in the documentation that different subcarriers are used for different streams. 
Here, the ``KroneckerPilotPattern`` is used which is a special case that allocates always full OFDM symbols for pilots and then "orthogonalizes" the pilot sequences for different streams.

Hope that helps.

**Reply by janfromme** [reply]
Hi Jakob,

thanks for the prompt reply! I tried that earlier but received an assertion basically stating that the number of "ones" in the `mask` is not equal to the number of pilots in the `pilots` tensor. That was confusing at first but now that I think about that (and also had a closer look at the code) it makes total sense.

I reduced the number of pilots in the `pilots` tensor according to the mask and now everything works as expected.

Thanks again!
Jan
